### ISSUE TITLE
Is ROX-Filer useable with GTK3 / Wayland?

### ISSUE THREAD

**Issue by woodenshoe-wi** [puppylinux-woof-CE/rox-filer#6]
I knew @dimkr had done some work to get ROX-Filer working with GTK3, but I didn’t expect it to work with Wayland because it uses X protocols to communicate with an existing instance when running any `rox` commands.

Apparently dimkr was able to get ROX-Filer to run under labwc / Xwayland, but getting it to display properly wouldn’t be trivial.
From https://github.com/puppylinux-woof-CE/woof-CE/issues/2617#issuecomment-963867290
> today I found out that when you run `rox -p` under labwc (and labwc is built with Xwayland support), the pinboard runs in a window. Pinboard icons work, and labwc manages all windows.

The labwc issue isn’t the appropriate place to discuss the future of ROX-Filer, but I am wondering what a Wayland based Puppy would look like.

Are there other options to implement a desktop/pinboard that already work well with Wayland?

Would a Wayland based Puppy not have desktop icons?

If using ROX-Filer, should the X communication protocols be replaced with dbus?  A named pipe?  Or something else?

I had thought with the move toward Wayland ROX-Filer might be getting to “end of life”, but maybe not?

**Comment by dimkr** [comment]
> it uses X protocols to communicate with an existing instance when running any `rox` commands.

Yes, and it also uses many deprecated GTK+ APIs gone in GTK+ 3.

Many things need to be removed from ROX-Filer to port it to GTK+ 3, including the pinboard.

And I don't think keeping it as-is is a good solution, because if it's going to be the only GTK+ 2 application in a Wayland desktop, we'll pay the extra RAM cost of Xwayland, and that would make ROX-Filer much heavier than many (most?) GTK+ 3 file managers.

> Would a Wayland based Puppy not have desktop icons?

I think we can (and should) implement a separate application for desktop icons - something friendlier and lighter than ROX-Filer.

Now that gtkdialog supports the layer shell protocol (https://github.com/puppylinux-woof-CE/gtkdialog/pull/102), technically, we can do it with an undecorated, fullscreen, gtkdialog window in the background layer 🤔 

> If using ROX-Filer, should the X communication protocols be replaced with dbus? A named pipe? Or something else?

Porting ROX-Filer to GTK+ 3 (and removing its own X dependencies) is a huge effort - I think the RPC thingy is only one part of the problem.

**Comment by woodenshoe-wi** [comment]
> Many things need to be removed from ROX-Filer to port it to GTK+ 3, including the pinboard.

So you were using the GTK+ 2 version with labwc/Xwayland and your efforts at  https://github.com/dimkr/rox-filer/tree/gtk3 didn’t work out?

Do you remember what the problem with GTK+ 3 and the pinboard was?  Is it impossible to put icons on the desktop with Wayland or is the problem that it requires a complete rewrite of ROX to use Wayland protocols?

Without the pinboard I don’t think it’s worth the work to migrate to GTK+ 3. ☹️

**Comment by dimkr** [comment]
> So you were using the GTK+ 2 version with labwc/Xwayland

Yes.

> your efforts at https://github.com/dimkr/rox-filer/tree/gtk3 didn’t work out?

I stopped working on this, at least for now. ROX-Filer uses deprecated GDK APIs (especially for the pinboard), and the porting effort is way more complex than I thought.

In addition, there's no "root window" in Wayland, so there's no place where ROX-Filer can draw the pinboard.

Technically, it should be possible to run ROX-Filer as a fullscreen window without decorations and put it beneath all other windows, but the background will be white, so it's not the same as today's pinboard.

**Comment by 01micko** [comment]
> but the background will be white, so it's not the same as today's pinboard.


Would this work in ~/.gtkrc-2.0.mine?

```
pixmap_path "/usr/share/backgrounds"
style "rox" = "default"
{
  bg_pixmap[NORMAL] = "Sky.svg"
}
widget_class "*Collection*" style "rox"
```

**Comment by dimkr** [comment]
@01micko Maybe, but we'll also need to figure out how to maximize the window, remove decorations, and move it to the background layer.

**Comment by woodenshoe-wi** [comment]
> In addition, there's no "root window" in Wayland, so there's no place where ROX-Filer can draw the pinboard.

Looking at the create_pinboard_window function in pinboard.c I don’t think ROX-Filer is drawing directly to the “root window”.

```
static void create_pinboard_window(Pinboard *pinboard)
...
win = gtk_window_new(GTK_WINDOW_TOPLEVEL);
...
gtk_widget_set_size_request(win, screen_width, screen_height);
...
gtk_window_move(GTK_WINDOW(win), 0, 0);
...
gdk_window_lower(win->window);
```

**Comment by 01micko** [comment]
I wonder if a wrapper program with `gtk-layer-shell` would work.

**Comment by woodenshoe-wi** [comment]
I guess I have the answers to my questions now.

The transition to GTK+ 3 needs the pinboard rewritten to use cairo, and currently ROX-Filer only runs in Wayland using Xwayland.

**Comment by dimkr** [comment]
@woodenshoe-wi Yes, and ROX-Filer has custom GTK+ widgets that don't compile against GTK+ 3. If I remember correctly - the icon view widget is one of them.

**Comment by 01micko** [comment]
This is interesting. I've been running the same test install for almost 2 months. It started out as a slacko64 install with cage (built from woof - ground up).

I ran the `labwc` petbuild installing directly. Failed to link statically with wlroots, but no big deal just backed out the `deps.patch` and all is running sweetly with Xwayland.

Take a look at this (the default.svg wallpaper shipped with the distro - I'm running dual screens)
![20211112_21h01m13s_grim](https://user-images.githubusercontent.com/1019119/141458364-bd246890-3405-4c98-97a4-b25698864459.png)

While it is true that rox will likely never run in a pure wayland setup, given that we have some tools to coerce rox with Xwayland then I believe that it could be possible to run the pinboard on a lower layer and everything could be quite happy.

I had not started `pup_event_fronted_d` (no use) but when I did start it then ran `rox -p xxx` to start the pinboard even the drive icons worked.

It _may_ not take a lot of effort to get rox to play nice with a wrapper. I could be wrong but it may be worth pursuing, for the purists at least.
